tst/PeerConnectionFunctionalityTest.cpp

```diff
@@ -106,6 +109,77 @@ TEST_F(PeerConnectionFunctionalityTest, connectTwoPeersForcedTURN)
+        return;
+    }
+
+    RtcConfiguration configuration;
+    PRtcPeerConnection offerPc = NULL, answerPc = NULL;
+    UINT32 i;
+    RtcSessionDescriptionInit sdp;
+    BOOL connected = FALSE;
+    SIZE_T stateChangeCount[7];
+
+    MEMSET(&configuration, 0x00, SIZEOF(RtcConfiguration));
+    MEMSET(stateChangeCount, 0x00, SIZEOF(stateChangeCount));
+
+    auto onICEConnectionStateChangeHdlr = [](UINT64 customData, RTC_PEER_CONNECTION_STATE newState) -> void {
+        ATOMIC_INCREMENT((PSIZE_T)customData + newState);
+    };
+
+    EXPECT_EQ(createPeerConnection(&configuration, &offerPc), STATUS_SUCCESS);
+    EXPECT_EQ(STATUS_SUCCESS, peerConnectionOnConnectionStateChange(offerPc,
+                                                                    (UINT64) stateChangeCount,
+                                                                    onICEConnectionStateChangeHdlr));
+    EXPECT_EQ(createPeerConnection(&configuration, &answerPc), STATUS_SUCCESS);
+    EXPECT_EQ(STATUS_SUCCESS, peerConnectionOnConnectionStateChange(answerPc,
+                                                                    (UINT64) stateChangeCount,
+                                                                    onICEConnectionStateChangeHdlr));
+
+    auto onICECandidateHdlr = [](UINT64 customData, PCHAR candidateStr) -> void {
+        if (candidateStr != NULL) {
+            std::thread([customData] (std::string candidate) {
+                RtcIceCandidateInit iceCandidate;
+                EXPECT_EQ(deserializeRtcIceCandidateInit((PCHAR) candidate.c_str(), STRLEN(candidate.c_str()), &iceCandidate), STATUS_SUCCESS);
+                EXPECT_EQ(addIceCandidate((PRtcPeerConnection) customData, iceCandidate.candidate), STATUS_SUCCESS);
+            }, std::string(candidateStr)).detach();
+        }
+    };
+
+    EXPECT_EQ(peerConnectionOnIceCandidate(offerPc, (UINT64) answerPc, onICECandidateHdlr), STATUS_SUCCESS);
+    EXPECT_EQ(peerConnectionOnIceCandidate(answerPc, (UINT64) offerPc, onICECandidateHdlr), STATUS_SUCCESS);
+
+    EXPECT_EQ(createOffer(offerPc, &sdp), STATUS_SUCCESS);
+    EXPECT_EQ(setLocalDescription(offerPc, &sdp), STATUS_SUCCESS);
+    EXPECT_EQ(setRemoteDescription(answerPc, &sdp), STATUS_SUCCESS);
+
+    EXPECT_EQ(createAnswer(answerPc, &sdp), STATUS_SUCCESS);
+    EXPECT_EQ(setLocalDescription(answerPc, &sdp), STATUS_SUCCESS);
+    EXPECT_EQ(setRemoteDescription(offerPc, &sdp), STATUS_SUCCESS);
+
+    for (i = 0; i < 10 && !connected; ++i) {
+        if (ATOMIC_LOAD(&stateChangeCount[RTC_PEER_CONNECTION_STATE_CONNECTED]) == 2) {
+            connected = TRUE;
+        } else {
+            THREAD_SLEEP(HUNDREDS_OF_NANOS_IN_A_SECOND);
+        }
+    }
+
+    EXPECT_TRUE(connected == TRUE);
+
+    // free offerPc so it wont send anymore keep alives and answerPc will detect disconnection
+    freePeerConnection(&offerPc);
+
+    THREAD_SLEEP((UINT64) (KVS_ICE_ENTER_STATE_DISCONNECTION_GRACE_PERIOD * 1.2));
```

Comment:
Is it possible this `1.2` can be racey? I would poll in a loop and do a higher time, in the good case we bail quickly but still avoid a possibly flakey test

Reply:
fixed